resolver/resolver.cc

```diff
@@ -556,46 +556,77 @@ class ResolveConstantsWalk {
-                e.setHeader("Must not pass your self to `{}`", send->fun.show(gs));
+            if (id->symbol.data(gs)->isClassOrModuleClass()) {
+                if (auto e =
+                        gs.beginError(core::Loc(todo.file, id->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
+                    e.setHeader("`{}` is a class, not a module; Only modules may be mixins",
+                                id->symbol.data(gs)->show(gs));
+                }
+                encounteredError = true;
             }
-            return;
-        }
-        auto existing = owner.data(gs)->findMember(gs, core::Names::classMethods());
-        if (existing.exists() && existing != id->symbol) {
-            if (auto e =
-                    gs.beginError(core::Loc(todo.file, send->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
-                e.setHeader("Redeclaring `{}` from module `{}` to module `{}`", send->fun.show(gs),
-                            existing.data(gs)->show(gs), id->symbol.data(gs)->show(gs));
+            if (id->symbol == owner) {
+                if (auto e =
+                        gs.beginError(core::Loc(todo.file, id->loc), core::errors::Resolver::InvalidMixinDeclaration)) {
+                    e.setHeader("Must not pass your self to `{}`", send->fun.show(gs));
+                }
+                encounteredError = true;
             }
-            return;
+            if (encounteredError) {
+                continue;
+            }
+
+            // Get the fake property holding the mixes
+            auto mixMethod = owner.data(gs)->findMember(gs, core::Names::mixedInClassMethods());
+            auto loc = core::Loc(owner.data(gs)->loc().file(), send->loc);
+            if (!mixMethod.exists()) {
+                // We never stored a mixin in this symbol
+                // Create a the fake property that will hold the mixed in modules
+                mixMethod = gs.enterMethodSymbol(loc, owner.asClassOrModuleRef(), core::Names::mixedInClassMethods());
+                vector<core::TypePtr> targs;
+                mixMethod.data(gs)->resultType = core::make_type<core::TupleType>(move(targs));
+
+                // Create a dummy block argument to satisfy sanitycheck during GlobalState::expandNames
+                auto &arg =
+                    gs.enterMethodArgumentSymbol(core::Loc::none(), mixMethod.asMethodRef(), core::Names::blkArg());
+                arg.flags.isBlock = true;
+            }
+
+            auto type = core::make_type<core::ClassType>(id->symbol.asClassOrModuleRef());
+            auto &elems = (core::cast_type<core::TupleType>(mixMethod.data(gs)->resultType))->elems;
+            // Make sure we are not adding existing symbols to our tuple
+            for (auto &elem : elems) {
+                if (type == elem) {
+                    continue;
+                }
+            }
+            elems.emplace_back(type);
```

Comment:
This continue will only apply to the loop defined on line 623. How about using `absl::c_find` instead? You might need to add an include for this to work.

```suggestion
            if (!absl::c_find(elems, type)) {
                elems.emplace_back(type);
            }
```

Reply:
Sorry I missed this. Added `if (absl::c_find(elems, type) == elems.end())` as `c_find` doesn't return a `bool`.